node/src/components/block_validator.rs

```diff
@@ -161,26 +192,284 @@ impl BlockValidator {
+        // before that, etc.) - these indices will correspond directly to the indices in
+        // RewardedSignatures.
+        past_blocks_with_metadata
+            .iter()
+            .rev()
+            .map(|maybe_metadata| {
+                maybe_metadata
+                    .as_ref()
+                    .map(|metadata| metadata.block.era_id())
+            })
+            .collect()
+    }
+
+    fn get_relevant_validators(
+        &mut self,
+        past_blocks_with_metadata: &[Option<BlockWithMetadata>],
+    ) -> HashMap<EraId, BTreeSet<PublicKey>> {
+        let era_ids_vec = Self::era_ids_vec(past_blocks_with_metadata);
+        // get the set of unique era ids that are present in the cited blocks
+        let era_ids: HashSet<_> = era_ids_vec.iter().flatten().copied().collect();
+        let validator_matrix = &self.validator_matrix;
+
+        era_ids
+            .into_iter()
+            .filter_map(move |era_id| {
+                validator_matrix
+                    .validator_weights(era_id)
+                    .map(|weights| (era_id, weights.into_validator_public_keys().collect()))
+            })
+            .collect()
+    }
+
+    fn handle_got_past_blocks_with_metadata<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        past_blocks_with_metadata: Vec<Option<BlockWithMetadata>>,
+        request: BlockValidationRequest,
+    ) -> Effects<Event>
+    where
+        REv: From<Event>
+            + From<FetcherRequest<Transaction>>
+            + From<FetcherRequest<FinalitySignature>>
+            + From<FatalAnnouncement>
+            + Send,
+    {
+        let rewarded_signatures = request.block.value().rewarded_signatures();
+
+        match Self::relevant_blocks_and_cited_signatures(
+            &past_blocks_with_metadata,
+            request.proposed_block_height,
+            rewarded_signatures,
+        ) {
+            Ok(blocks_and_signatures) => {
+                let validators = self.get_relevant_validators(&past_blocks_with_metadata);
+
+                // This will be a set of signature IDs of the signatures included in the block, but
+                // not found in metadata in storage.
+                let mut missing_sigs = HashSet::new();
+
+                for (block_with_metadata, single_block_rewarded_sigs) in blocks_and_signatures {
+                    let era_id = block_with_metadata.block.era_id();
+                    let Some(all_validators) = validators.get(&era_id) else {
+                        return fatal!(effect_builder, "couldn't get validators for {}", era_id).ignore();
+                    };
+                    let public_keys = single_block_rewarded_sigs
+                        .clone()
+                        .to_validator_set(all_validators.iter().cloned());
+                    let block_hash = *block_with_metadata.block.hash();
+                    let era_id = block_with_metadata.block.era_id();
```

Comment:
You already have this binding a few lines above

Reply:
Good catch!